dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1390 @@
+        double[:] prod = np.ndarray(shape=(n,), dtype=np.float64)
+        double[:] x = np.ndarray(shape=(2,), dtype=np.float64)
+
+    grad_pdf[...] = 0
+    with nogil:
+        valid_points = 0
+        for i in range(nrows):
+            for j in range(ncols):
+                if smask is not None and smask[i, j] == 0:
+                    continue
+                if mmask is not None and mmask[i, j] == 0:
+                    continue
+
+                valid_points += 1
+                x[0] = _apply_affine_2d_x0(i, j, 1, grid_to_space)
+                x[1] = _apply_affine_2d_x1(i, j, 1, grid_to_space)
+
+                if constant_jacobian == 0:
+                    constant_jacobian = transform._jacobian(theta, x, J)
+
+                for k in range(n):
+                    prod[k] = J[0,k] * mgradient[i,j,0] +\
+                              J[1,k] * mgradient[i,j,1]
+
+                rn = _bin_normalize(static[i, j], smin, sdelta)
+                r = _bin_index(rn, nbins, padding)
+                cn = _bin_normalize(moving[i, j], mmin, mdelta)
+                c = _bin_index(cn, nbins, padding)
+                spline_arg = (c-2) - cn
+
+                for offset in range(-2, 3):
+                    val = _cubic_spline_derivative(spline_arg)
+                    for k in range(n):
+                        grad_pdf[r, c + offset,k] -= val * prod[k]
+                    spline_arg += 1.0
+
+        if valid_points * mdelta > 0:
+            for i in range(nbins):
+                for j in range(nbins):
+                    for k in range(n):
+                        grad_pdf[i, j, k] /= (valid_points * mdelta)
+
+
+cdef _joint_pdf_gradient_dense_3d(double[:] theta, Transform transform,
+                                  double[:,:,:] static, double[:,:,:] moving,
+                                  double[:,:] grid_to_space,
+                                  floating[:,:,:,:] mgradient, int[:,:,:] smask,
+                                  int[:,:,:] mmask, double smin, double sdelta,
+                                  double mmin, double mdelta, int nbins,
+                                  int padding, double[:,:,:] grad_pdf):
+    r''' Gradient of the joint PDF w.r.t. transform parameters theta
+
+    Computes the vector of partial derivatives of the joint histogram w.r.t.
+    each transformation parameter. The transformation itself is not necessary
+    to compute the gradient, but only its Jacobian.
+
+    Parameters
+    ----------
+    theta: array, shape (n,)
+        parameters of the transformation to compute the gradient from
+    transform: instance of Transform
+        the transformation with respect to whose parameters the gradient
+        must be computed
+    static: array, shape (S, R, C)
+        static image
+    moving: array, shape (S, R, C)
+        moving image
+    grid_to_space: array, shape (4, 4)
+        the grid-to-space transform associated to images static and moving
```

Comment:
That's right. Regarding the naming, I was thinking that voxel_to_mm maybe ambiguous too: With `voxel_to_mm` do we mean 
a) The diagonal matrix that scales from voxel units to mm units (by multiplying by the voxel size)?
or
b) The affine transform mapping grid coordinates to points in physical space?
In most cases we will mean b) but I wonder if we can come up with a better naming convention